### ISSUE TITLE
RestQL Server does not found services by name on SwarmMode

### ISSUE THREAD

**Issue by rafapg** [B2W-BIT/restQL-server#13]
I'm trying to use the RestQL Server to pull information from my microservices at the same docker swarm infrastructure that the RestQL Server runs. Unfortunately, the server can't find any microservice.

When I configure the RestQL Server to use the resources outside this docker infrastructure, it seems to work fine.

When running everything on docker, if I access the container of the server and run a web request (with `cURL`) to a  microservice (using the service name) the request occurs successfully. So, the issue appears to be related with the JVM, the Clojure implementation or the JVM DNS resolution name inside a docker container.

We also tried run using `docker-compose` and got the same error.

Thanks in advance.
 
UPDATE: The RestQL can access my mongo  instance (RestQL Quey Database) running in the docker infrastructure.

**Comment by ricardoekm** [comment]
Hi @rafapg,
Can you please post the exact restQL query response?

Is there any error in the logs or in the stdout?

**Comment by Kasama** [comment]
Hello @ricardoekm, I work with @rafapg and can provide some more information on the problem.

Here is a minimal docker-compose.yml, based on the one in this repository, to reproduce the problem:
```
version: '3'
services:
  restql:
    build: .
    image: restql-server:alpine
    restart: always
    environment:
      - planets=http://swapi.co/api/planets/
      - squares=http://square-api:3000/square/:number/
    links:
      - square-api
      - mongodb
    ports:
      - "0.0.0.0:9000:9000"

  square-api:
    image: robertoaall/simple-square-api
    ports:
      - "3000:3000"

  mongodb:
    image: mongo:3.4
    network_mode: host

  nginx:
    image: nginx:1.12-alpine
    volumes:
      - ./docker/nginx/:/etc/nginx/conf.d/
    links:
      - restql
    ports:
      - 8080:80
```
The RestQL application works as expected with the `planets` resource
```
curl -H "Content-Type: text/plain" localhost:8080/run-query -d "from planets with name = \"earth\""

#=> {"planets":{"details":{"success":true,"status":200,"metadata":{}},"result":{...}}}
```

But a request to the `squares` resouce returns
```
curl -H "Content-Type: text/plain" localhost:8080/run-query -d "from squares with number = 2"

#=> {"message":{"error":"VALIDATION_ERROR","message":":from as a keyword should reference a valid mapped resource. Error was in :squares"}}
```
and the RestQL logs print out:
```
restql_1      | Feb 09, 2018 1:55:25 AM restql.core.log invoke
restql_1      | SEVERE: (({:time 36, :success false} restQL Query finished))
nginx_1       | 172.21.0.1 - - [09/Feb/2018:01:58:10 +0000] "POST /run-query HTTP/1.1" 200 5445 "-" "curl/7.57.0" "-"
```
---
if I `docker exec -it restqlserver_restql bash` into the container and install `curl`, it's possible to request the resource as expected:
```
curl http://square-api:3000/square/2/

#=> {"result":4}
```

**Comment by Kasama** [comment]
We've also noticed that if the RestQL server is put into `host` network mode, it's possible to access the the resource using the `localhost` URL.

Example docker-compose.yml (nginx was removed for simplicity)
```
version: '3'
services:
  restql:
    build: .
    image: restql-server:alpine
    restart: always
    environment:
      - planets=http://swapi.co/api/planets/
      - squares=http://localhost:3000/square/:number/
    network_mode: host
    ports:
      - "0.0.0.0:9000:9000"

  square-api:
    image: robertoaall/simple-square-api
    ports:
      - "3000:3000"

  mongodb:
    image: mongo:3.4
    network_mode: host
```
then the request works as expected
```
curl -H "Content-Type: text/plain" localhost:9000/run-query -d "from squares with number = 2"

#=> {"squares":{"details":{"success":true,"status":200,"metadata":{}},"result":{"result":4}}}
```

**Comment by sio-iago** [comment]
Hi Kasama,

This mapping error probably occurs because we’re using Apache URL commons to validade if a mapped url is a valid URL. This gets skipped if the url is localhost (hardcoded).

We could have two solutions here:

1. You could try to nap this service to a valid URL (according to the lib, any DNS valid URL) such as squares.api/

2. We can consider removing this URL validation for sake of simplicity, but it needs to be aligned accordingly with the team to make sure we don’t create any potential security flaws.

Could you please try 1. And tell us if it works, so we can confirm this and patch the solution 2?

**Comment by ricardoekm** [comment]
Yep, its weird apache URL commons doesn't consider valid URL without TDL (e.g. http://intranet/).

I think we should patch it to allow such URLs as they might be common on internal calls. I'll check it and submit a pull request.

**Comment by Kasama** [comment]
thanks @sio-iago, I've changed the docker-compose.yml so that the RestQL service used the name `square.api` as suggested:
```
  restql:
    build: .
    image: restql-server:alpine
    restart: always
    environment:
      - planets=http://swapi.co/api/planets/
      - squares=http://square.api:3000/square/:number/
    links:
      - square-api:square.api
      - mongodb
    ports:
      - "0.0.0.0:9000:9000"
```

and the request worked as expected.
---
May I suggest another method to validate the URL, such as a ping or a dummy request during application boot.

**Comment by sio-iago** [comment]
We fixed this issue and it will be released soon. Refer to [this commit](https://github.com/B2W-BIT/restQL-core/commit/09807bd8cbbec88339cdb3e8ef30407648069615) for more information.

**Comment by sio-iago** [comment]
Tagged to be part of release 2.3.1

**Comment by ricardoekm** [comment]
We've just released the version 2.3.1. Can you please give it try?

Please let us know if you need further assistance, don't hesitate to open other issues.

**Comment by Kasama** [comment]
Thanks, it's working as expected now. Both in the minimal example and in our stack. 👍
